CC one electron properties logic

amjames opened the pull request:
## Description
Modifies how one electron properties are computed to allow the `properties` kwarg to behave consistently with other modules.

## Todos
Notable points that this PR has either accomplished or will accomplish.
* **Developer Interest**
  - [x] CCDENSITY no longer calls into OEPROPs this is now done in the python layer. 
  - [x] After calling CCDENSITY the `wfn.arrays['CC ROOT {n} D{a|b}']` can be used to access the OPDM for excited states in the SO basis as in the convention for matrix quantities set on the wfn. 
* **User-Facing for Release Notes**
  - [x] Excited state OPDMs are now set in the wfn.arrays, so that they can be used to compute 1e properties for all roots requested in EOM-CC* calls.
 - [x] Now additional valid one electron properties can be passed as part of the `properties=[]` list. These would previously cause a validation error. Valid one electron properties are now:
    - dipole *
    - quadrupole *
    - mulliken_charges **
    - lowdin_charges
    - mayer_indices
    - wiberg_lowdin_indices
    - no_occupations **
  - [x] The previous behavior was such that requesting the properties marked with * above would lead to all * -ed properties and additional (**) properties being computed automatically. The new logic implemented in the driver still has this behavior so existing workflows do no need to change, at least for now. 

## Questions
- [x] @loriab The block of code that I removed from CCDENSITY had comments that were used to trigger documentation of some psivars. Where should these go now? 

## Checklist
- [x] Tests added for any new features
- [x] [All or relevant fraction of full tests run](http://psicode.org/psi4manual/master/build_planning.html#how-to-run-a-subset-of-tests)
  -  I ran all of the CC tests and there were no errors. The new logic changes where in the output file the 1 electron properties appear, but the perl script still finds them since they are labeled/formatted the same way. 

## Status
- [x] Ready for review
- [x] Ready for merge

loriab commented on psi4/driver/procrouting/proc.py:
```diff
@@ -2551,6 +2548,12 @@ def run_cc_property(name, **kwargs):
+    if ('dipole' in one) or ('quadrupole' in one):
+        for extra_1e in ['dipole', 'quadrupole', 'mulliken_charges', 'no_occupations']:
+            if extra_1e not in one:
+                one.append(extra_1e)
```
Is this the same logic?
```
if ('dipole' in one) or ('quadrupole' in one):
    one = list(set(one.extend(['dipole', 'quadrupole', 'mulliken_charges', 'no_occupations'])))
```

loriab commented on psi4/src/psi4/ccdensity/ccdensity.cc:
```diff
@@ -383,72 +383,29 @@ PsiReturnType ccdensity(std::shared_ptr<Wavefunction> ref_wfn, Options& options)
-        Process::environment.globals["CC ROOT 0 QUADRUPOLE XZ"] = Process::environment.globals["CC QUADRUPOLE XZ"];
-        Process::environment.globals["CC ROOT 0 QUADRUPOLE YY"] = Process::environment.globals["CC QUADRUPOLE YY"];
-        Process::environment.globals["CC ROOT 0 QUADRUPOLE YZ"] = Process::environment.globals["CC QUADRUPOLE YZ"];
-        Process::environment.globals["CC ROOT 0 QUADRUPOLE ZZ"] = Process::environment.globals["CC QUADRUPOLE ZZ"];
```
Just leave the P::e block here, only commented. psivar scraper for docs is so dumb it doesn't notice.

loriab commented on psi4/src/psi4/ccdensity/ccdensity.cc:
```diff
@@ -383,72 +383,29 @@ PsiReturnType ccdensity(std::shared_ptr<Wavefunction> ref_wfn, Options& options)
-      /*- Process::environment.globals["CC ROOT n QUADRUPOLE XZ"] -*/
-      /*- Process::environment.globals["CC ROOT n QUADRUPOLE YY"] -*/
-      /*- Process::environment.globals["CC ROOT n QUADRUPOLE YZ"] -*/
-      /*- Process::environment.globals["CC ROOT n QUADRUPOLE ZZ"] -*/
```
Same here wrt P::e psivars.

dgasmith commented on psi4/driver/procrouting/proc.py:
```diff
@@ -2514,16 +2514,13 @@ def run_cc_property(name, **kwargs):
-        ['CCLAMBDA', 'R_CONVERGENCE'],
-        ['CCEOM', 'R_CONVERGENCE'],
-        ['CCEOM', 'E_CONVERGENCE'])
+    optstash = p4util.OptionsState(['WFN'], ['DERTYPE'], ['ONEPDM'], ['PROPERTY'], ['CCLAMBDA', 'R_CONVERGENCE'],
```
For whatever reason, we usually don't format this as I think the other way seems cleaner.

amjames replied:
Should I just block it off with `yapf: enable/disable`?

loriab replied:
add `# yapf: disable` at the end of the line so it's self-contained.

dgasmith commented on psi4/driver/procrouting/proc.py:
```diff
@@ -2514,16 +2514,13 @@ def run_cc_property(name, **kwargs):
+    oneel_properties = [
+        'dipole', 'quadrupole', 'mulliken_charges', 'lowdin_charges', 'mayer_indices', 'wiberg_lowdin_indices',
+        'no_occupations'
+    ]
```
You might want to check the other functions but OEProp should carry around its valid list. It would be good to pull from there.

amjames replied:
Good point. I thought that list would include transition dipole/quadrupole but I should have checked.

amjames commented:
I fixed the `oneel_properties` issues, so that the list is not hard-coded in there any more. I also ran `ccdensity.cc` through `c-f` and I removed some includes that are not needed anymore.

amjames commented:
Yep, merge this at your leisure!
